com.unity.perception/Runtime/Randomization/Randomizers/AssetSources/AssetSource.cs

```diff
@@ -0,0 +1,139 @@
+
+        bool m_Initialized;
+        UniformSampler m_Sampler = new UniformSampler();
+
+        /// <summary>
+        /// The archetype used to preprocess assets from this source
+        /// </summary>
+        public Archetype<T> archetype
+        {
+            get => (Archetype<T>)m_ArchetypeBase;
+            set => m_ArchetypeBase = value;
+        }
+
+        /// <summary>
+        /// The number of assets available within this asset source
+        /// </summary>
+        public int Count => assetSourceLocation.Count;
+
+        /// <summary>
+        /// Execute setup steps for this AssetSource
+        /// </summary>
+        public void Initialize()
+        {
+            assetSourceLocation.Initialize(archetype);
+            m_Initialized = true;
+        }
+
+        /// <summary>
+        /// Returns the unprocessed asset loaded from the provided index
+        /// </summary>
+        /// <param name="index">The index of the asset to load</param>
+        /// <returns>The asset loaded at the provided index</returns>
+        public T LoadRawAsset(int index)
+        {
+            CheckIfInitialized();
+            return assetSourceLocation.LoadAsset<T>(index);
+        }
+
+        /// <summary>
+        /// Returns all unprocessed assets that can be loaded from this AssetSource
+        /// </summary>
+        /// <returns>All assets that can be loaded from this AssetSource</returns>
+        public T[] LoadAllRawAssets()
+        {
+            var array = new T[Count];
+            for (var i = 0; i < Count; i++)
+            {
+                array[i] = LoadRawAsset(i);
+                archetype.Preprocess(array[i]);
+            }
+            return array;
+        }
+
+        /// <summary>
+        /// Creates an instance of the asset loaded from the provided index and preprocesses it using the archetype
+        /// assigned to this asset source
+        /// </summary>
+        /// <param name="index">The index of the asset to load</param>
+        /// <returns>The instantiated instance</returns>
+        public T CreateProcessedInstance(int index)
+        {
+            CheckIfInitialized();
+            return CreateProcessedInstance(LoadRawAsset(index));
+        }
+
+        /// <summary>
+        /// Instantiates, preprocesses, and returns all assets that can be loaded from this asset source
+        /// </summary>
+        /// <returns>Instantiated instances from every loadable asset</returns>
+        public T[] CreateProcessedInstances()
+        {
+            var array = new T[Count];
+            for (var i = 0; i < Count; i++)
+                array[i] = CreateProcessedInstance(i);
+            return array;
+        }
+
+        /// <summary>
+        /// Returns a uniformly random sampled asset from this AssetSource
+        /// </summary>
+        /// <returns>The randomly sampled asset</returns>
+        public T SampleAsset()
```

Comment:
Why randomize in this class? Randomization logic appears to be outside the scope of this class.

Reply:
I'm down to remove the sampling features, but I just felt that simple uniform sampling would be a common use case, similar to how users sample from GameObjectParameters.

Reply:
I've removed the randomization features